spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/MqttAdapterTests.java

```diff
@@ -369,6 +377,34 @@ public void testCustomExpressions() {
 		ctx.close();
 	}
 
+	@Test
+	public void testReconnect() throws Exception {
+		final MqttAsyncClient client = mock(MqttAsyncClient.class);
+		MqttPahoMessageDrivenChannelAdapter adapter = buildAdapter(client, null, ConsumerStopAction.UNSUBSCRIBE_NEVER);
+		adapter.setRecoveryInterval(10);
+		Log logger = spy(TestUtils.getPropertyValue(adapter, "logger", Log.class));
+		new DirectFieldAccessor(adapter).setPropertyValue("logger", logger);
+		given(logger.isDebugEnabled()).willReturn(true);
+		final AtomicInteger attemptingReconnectCount = new AtomicInteger();
+		willAnswer(i -> {
+			if (attemptingReconnectCount.getAndIncrement() == 0) {
+				adapter.connectionLost(new RuntimeException("while schedule running"));
+			}
+			i.callRealMethod();
+			return null;
+		}).given(logger).debug("Attempting reconnect");
+		ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
+		taskScheduler.initialize();
+		adapter.setTaskScheduler(taskScheduler);
+		adapter.start();
+		adapter.connectionLost(new RuntimeException("initial"));
+		Thread.sleep(1000);
```

Comment:
Works for me even without this.
Am I missing anything?
Can't we switch to the `CountDownLatch` in the `willAnswer` impl? 

Aren't we pursuing the goal to be sure that reconnect called several times in case of `connectionLost()`?

Or we can `spy()` on the `TaskScheduler.schedule()` for the same `latch` reason